tests/test_uri.py

```diff
@@ -0,0 +1,330 @@
+
+    def testFile(self):
+        file = os.path.join(self.tmpdir, "test.txt")
+        uri = ButlerURI(file)
+        self.assertFalse(uri.exists(), f"{uri} should not exist")
+        self.assertEqual(uri.ospath, file)
+
+        content = "abcdefghijklmnopqrstuv\n"
+        uri.write(content.encode())
+        self.assertTrue(os.path.exists(file), "File should exist locally")
+        self.assertTrue(uri.exists(), f"{uri} should now exist")
+        self.assertEqual(uri.read().decode(), content)
+
+    def testRelative(self):
+        """Check that we can get subpaths back from two URIs"""
+        parent = ButlerURI(self.tmpdir, forceDirectory=True, forceAbsolute=True)
+        child = ButlerURI(os.path.join(self.tmpdir, "dir1", "file.txt"), forceAbsolute=True)
+
+        self.assertEqual(child.relative_to(parent), "dir1/file.txt")
+
+        not_child = ButlerURI("/a/b/dir1/file.txt")
+        self.assertFalse(not_child.relative_to(parent))
+
+        not_directory = ButlerURI(os.path.join(self.tmpdir, "dir1", "file2.txt"))
+        self.assertFalse(child.relative_to(not_directory))
+
+        # Relative URIs
+        parent = ButlerURI("a/b/", forceAbsolute=False)
+        child = ButlerURI("a/b/c/d.txt", forceAbsolute=False)
+        self.assertFalse(child.scheme)
+        self.assertEqual(child.relative_to(parent), "c/d.txt")
+
+        # File URI and schemeless URI
+        parent = ButlerURI("file:/a/b/c/")
+        child = ButlerURI("e/f/g.txt", forceAbsolute=False)
+
+        # If the child is relative and the parent is absolute we assume
+        # that the child is a child of the parent unless it uses ".."
+        self.assertEqual(child.relative_to(parent), "e/f/g.txt")
+
+        child = ButlerURI("../e/f/g.txt", forceAbsolute=False)
+        self.assertFalse(child.relative_to(parent))
+
+        child = ButlerURI("../c/e/f/g.txt", forceAbsolute=False)
+        self.assertEqual(child.relative_to(parent), "e/f/g.txt")
+
+    def testMkdir(self):
+        tmpdir = ButlerURI(self.tmpdir)
+        newdir = tmpdir.join("newdir/seconddir")
+        newdir.mkdir()
+        self.assertTrue(newdir.exists())
+        newfile = newdir.join("temp.txt")
+        newfile.write("Data".encode())
+        self.assertTrue(newfile.exists())
+
+    def testTransfer(self):
+        src = ButlerURI(os.path.join(self.tmpdir, "test.txt"))
+        content = "Content is some content\nwith something to say\n\n"
+        src.write(content.encode())
+
+        for mode in ("copy", "link", "hardlink", "symlink", "relsymlink"):
+            dest = ButlerURI(os.path.join(self.tmpdir, f"dest_{mode}.txt"))
+            dest.transfer_from(src, transfer=mode)
+            self.assertTrue(dest.exists(), f"Check that {dest} exists (transfer={mode})")
+
+            with open(dest.ospath, "r") as fh:
+                new_content = fh.read()
+            self.assertEqual(new_content, content)
+
+            if "symlink" in mode:
```

Comment:
Also `relsymlink` too, yes? It would probably also be worth checking the value of the symlink.

Reply:
Yes, that's why I used `in` and not `==` (since relsymlink does include symlink).

Reply:
Oh, right, Python `in` works with strings as well as proper collections. I always find this confusing. To be honest, I personally would still write this as `if mode in ("symlink", "relsymlink")` because I think it's more obvious and easier to read, but that's a very minor nit.